modules/core/04-channel/keeper/upgrade.go

```diff
@@ -302,6 +302,132 @@ func (k Keeper) WriteUpgradeAckChannel(
+	}
+
+	if connection.GetState() != int32(connectiontypes.OPEN) {
+		return errorsmod.Wrapf(connectiontypes.ErrInvalidConnectionState, "connection state is not OPEN (got %s)", connectiontypes.State(connection.GetState()).String())
+	}
+
+	var counterpartyChannel types.Channel
+	switch counterpartyChannelState {
+	case types.OPEN:
+		upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+		if !found {
+			return errorsmod.Wrapf(types.ErrUpgradeNotFound, "failed to retrieve channel upgrade: port ID (%s) channel ID (%s)", portID, channelID)
+		}
+
+		counterpartyChannel = types.Channel{
+			State:           types.OPEN,
+			Ordering:        upgrade.Fields.Ordering,
+			ConnectionHops:  upgrade.Fields.ConnectionHops,
+			Counterparty:    types.NewCounterparty(portID, channelID),
+			Version:         upgrade.Fields.Version,
+			UpgradeSequence: channel.UpgradeSequence,
+			FlushStatus:     types.NOTINFLUSH,
+		}
+
+	case types.TRYUPGRADE:
+		// If the counterparty is in TRYUPGRADE, then we must have gone through the ACKUPGRADE step.
+		if channel.State != types.ACKUPGRADE {
+			return errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.ACKUPGRADE, channel.State)
+		}
+
+		counterpartyChannel = types.Channel{
+			State:           types.TRYUPGRADE,
+			Ordering:        channel.Ordering,
+			ConnectionHops:  []string{connection.GetCounterparty().GetConnectionID()},
+			Counterparty:    types.NewCounterparty(portID, channelID),
+			Version:         channel.Version,
+			UpgradeSequence: channel.UpgradeSequence,
+			FlushStatus:     types.FLUSHCOMPLETE,
+		}
+
+	case types.ACKUPGRADE:
+		counterpartyChannel = types.Channel{
+			State:           types.ACKUPGRADE,
+			Ordering:        channel.Ordering,
+			ConnectionHops:  []string{connection.GetCounterparty().GetConnectionID()},
+			Counterparty:    types.NewCounterparty(portID, channelID),
+			Version:         channel.Version,
+			UpgradeSequence: channel.UpgradeSequence,
+			FlushStatus:     types.FLUSHCOMPLETE,
+		}
+
+	default:
+		panic(fmt.Sprintf("counterparty channel state should be in one of [%s, %s, %s]; got %s", types.TRYUPGRADE, types.ACKUPGRADE, types.OPEN, counterpartyChannelState))
+	}
+
+	err = k.connectionKeeper.VerifyChannelState(ctx, connection, proofHeight, proofChannel, portID, channelID, counterpartyChannel)
+	if err != nil {
+		return errorsmod.Wrapf(err, "failed to verify counterparty channel, expected counterparty channel state: %s", counterpartyChannel.String())
+	}
+
+	return nil
+}
+
+// WriteUpgradeOpenChannel writes the agreed upon upgrade fields to the channel, sets the channel flush status to NOTINFLUSH and sets the channel state back to OPEN. This can be called in one of two cases:
+// - In the UpgradeAck step of the handshake if both sides have already flushed all in-flight packets.
+// - In the UpgradeOpen step of the handshake.
+func (k Keeper) writeUpgradeOpenChannel(ctx sdk.Context, portID, channelID string) {
```

Comment:
Yeah, guess the idea for Ack was test since there's the conditional there that leads to two different states. I do think we're lacking test cov on some checks in the write functions for the other handlers so I'd definitely be in favour of testing the rest in a follow up.